src/HerbConstraints.jl

```diff
@@ -15,6 +16,85 @@ Grammar constraints should implement `on_new_node` to post a [`AbstractLocalCons
 """
 abstract type AbstractGrammarConstraint <: AbstractConstraint end
 
+# Interface for update_rule_indices!
+"""
+    function update_rule_indices!(node::AbstractGrammarConstraint,n_rules::Integer)
+
+Used to update constraints when the grammar changes, e.g., by adding rules or merging grammars.
+
+# Arguments
+- `c`: The `AbstractGrammarConstraint` to be updated
+- `n_rules`: The new number of rules in the grammar
+
+# Notes
+Individual implementations may not use all provided arguments.
+"""
+function HerbCore.update_rule_indices!(c::AbstractGrammarConstraint,
+    n_rules::Integer
+)
+    error("update_rule_indices! not implemented for constraint type $(typeof(c))")
+end
+
+"""
+    function update_rule_indices!(node::AbstractGrammarConstraint,grammar::AbstractGrammar)
+
+Used to update constraints when the grammar changes, e.g., by adding rules or merging grammars.
+
+# Arguments
+- `c`: The `AbstractGrammarConstraint` to be updated
+- `grammar`: The grammar that changed
+
+# Notes
+Individual implementations may not use all provided arguments.
+"""
+function HerbCore.update_rule_indices!(c::AbstractGrammarConstraint,
+    grammar::AbstractGrammar
+)
+    error("update_rule_indices! not implemented for constraint type $(typeof(c))")
```

Comment:
Same as the other PR in Core (https://github.com/Herb-AI/HerbCore.jl/pull/53#discussion_r2129083488), I would prefer we don't implement these methods.